packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -200,6 +200,13 @@ class ImageCache {
     _currentSizeBytes = 0;
   }
 
+  /// All keys inside the current cache.
+  Set<Object> get keys => <Object>{
+    ..._liveImages.keys,
+    ..._pendingImages.keys,
+    ..._cache.keys,
+  };
```

Comment:
I'd rather not expose this API. Partly because it's a getter that does a lot of work,and partly because it exposes lots of objects to external classes that they shouldn't have access to.

Can we instead make the `key` object on `NetworkImage` return the right key?

Reply:
> Can we instead make the key object on NetworkImage return the right key?

I'm afraid it is not very easy without breaking changes.

Currently, ResizeImage:

```dart
  ImageStreamCompleter load(ResizeImageKey key, DecoderCallback decode) {
...
    imageProvider.load(key._providerCacheKey, decodeResize);
```

and NetworkImage:

```dart
  ImageStreamCompleter load(image_provider.NetworkImage key, image_provider.DecoderCallback decode) {
...
```

Given that, a ResizeImage can be combined with *any* other image providers (e.g. the popular CachedNetworkImage in https://pub.dev/packages/cached_network_image which I use), I guess we should not change the API of `load` function.

Reply:
But this fix won't help other wrapping image providers inthe framework or out in the wild either. We shouldn't have NetworkImage know anything about ResizeImage to fix this bug.

Reply:
For example, perhaps ResizeImage should listen to the completer returned by its wrapped provider and evict itself if the wrapped provider's `ImageStreamCompleter` calls the listener with `onError`.

I'm not sure if we'd need some kind of error-only listening API to make sure we don't dispose the stream when the resize completer stops listening to it.

Reply:
> But this fix won't help other wrapping image providers inthe framework or out in the wild either. We shouldn't have NetworkImage know anything about ResizeImage to fix this bug.

I agree, I also do not like the current fix :(

> For example, perhaps ResizeImage should listen to the completer returned by its wrapped provider and evict itself if the wrapped provider's ImageStreamCompleter calls the listener with onError.

Good idea!

> I'm not sure if we'd need some kind of error-only listening API to make sure we don't dispose the stream when the resize completer stops listening to it.

I will try and see.

Reply:
But here is a problem: When shall we dispose the listener if we call ImageStreamCompleter addListener inside ResizedImage?

I checked ImageProvider's API, and there is no dispose, so we cannot call it there.

In addition, if the image is a infinitely-playing gif, we cannot simply removeListener when seeing the first frame of image, since there are infinitely many frames.

Reply:
Another possibility: Breaking change the `loadImage(key)`, into something like `loadImage(keyToLoadImage, keyThatIsUsedInImageCache)`. Then,

```dart
class ResizedImage {
  loadImage(keyToLoadImage, keyThatIsUsedInImageCache, ...) {
    imageProvider.load(keyToLoadImage: key._providerCacheKey, keyThatIsUsedInImageCache: key, decodeResize);
  }
}

class NetworkIMage {
  loadImage(keyToLoadImage, keyThatIsUsedInImageCache, ...) {
    // Use keyToLoadImage like the old code, except that we use keyThatIsUsedInImageCache to call `ImageCache.evict`
  }
}
```

Reply:
I think if there's been at least one frame we can remove the listener. Evicting a multi-frame image because one frame causes an exception isn't really the main concern here, it's about if the provider itself fails to acquire the data.

Reply:
I see. I will try to implement that and see!

Reply:
Implemented, and CI is green now

![image](https://github.com/flutter/flutter/assets/5236035/5ab72005-f7e3-4e59-8146-8908028a3fbd)